File: 592-fraction-addition-and-subtraction/592-fraction-addition-and-subtraction.cpp
class Solution {
public:
string fractionAddition(string expression) {
    istringstream in(expression);
    char chr;

    int A = 0, B = 1, a, b;

    while (in >> a >> chr >> b) {
        A = A * b + a * B;
        B *= b;
        int g = abs(__gcd(A, B));
        A /= g;
        B /= g;
    }
    return to_string(A) + '/' + to_string(B);
}
};

File: 592-fraction-addition-and-subtraction/592-fraction-addition-and-subtraction.java
class Solution {
  public String fractionAddition(String expression) {
    int first = 0;
    int second = 1;
    
    Scanner sc = new Scanner(expression).useDelimiter("/|(?=[+-])");

    for (;sc.hasNext();) {
      final int valA = sc.nextInt();
      final int valB = sc.nextInt();
      first = first * valB + valA * second;
      second *= valB;
      final int g = gcd(first, second);
      first /= g;
      second /= g;
    }

    return first + "/" + second;
  }

  private int gcd(int num1, int num2) {
    return num1 == 0 ? Math.abs(num2) : gcd(num2 % num1, num1);
  }
}
